### PR TITLE
Added login auth support and HTML email support, enabled annotations, subject templating

### LINKED ISSUE #18: Error: failed to send email: 504 5.7.4 Unrecognized authentication type
Hi,
I'm trying to send emails using `smtp.office365.com` but I get this error:

`Error: failed to send email: 504 5.7.4 Unrecognized authentication type [DBBPR09CA0015.eurprd09.prod.outlook.com]`

I'm using the same credential/settings successfully used by other applications (e.g. grafana).

Thanks

### LINKED ISSUE #20: Is it able to parse a html or erb templates?
Hi Team,

I have template as mentioned below
Sensu Alert - {{.Entity.Name}}

I am able to parse above template and I have received email

Is this script able to parse html or erb template as mentioned below
<html>
<head>
</style>
</head>
<body>
        <tr >
                <td>Check Name: </td>
                <td> {{.Check.Name}}
                </td>
        </tr>
</body>
</html>

As I checked, I am not able to parse it 
How Can I present the mail in a tabular format

### PR DESCRIPTION
**Pull request by nixwiz**
Closes #18 
Closes #20 

Also adds support for managing the subject via a template.

### CODE REVIEW

**Comment by aphix** [comment]
Any chance we can get this merged and released?

I'm really looking to get auth support configured so I can have a working email handler.

Thanks for your work @nixwiz!

**Comment by nixwiz** [comment]
I'm at the mercy of the Sensu staff here @aphix.  My original pull request sat untouched for months only to be more or less bulldozed by other changes.  I had to refactor my updates and resubmit and here we sit again, waiting.

Maybe @nikkiki can provide some insight as to why.

**Comment by portertech** [comment]
Hey @nixwiz! Sorry for dropping the ball on your PR (again) 😓 These changes look great and I do not believe they conflict with other work nor plans.

@nikkiki WDYT?

File: main.go [hunk 1]
```diff
@@ -38,20 +44,22 @@ const (
 	fromEmail        = "fromEmail"
 	insecure         = "insecure"
 	hookout          = "hookout"
+	loginauth        = "loginauth"
```
Comment (nikkiki) [inline main.go]:
```suggestion
	loginAuth        = "loginAuth"
```

**Comment by portertech** [comment]
@nixwiz I assume login is working for you?

**Comment by nixwiz** [comment]
@portertech actually, @aphix had tested the changes on my prior PR.

File: main.go [hunk 2]
```diff
@@ -234,3 +272,30 @@ func resolveTemplate(templateValue string, event *corev2.Event) (string, error)
...
+
+func (a *loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
+	if more {
+		switch string(fromServer) {
+		case "Username:":
+			return []byte(a.username), nil
+		case "Password:":
+			return []byte(a.password), nil
+		default:
+			return nil, errors.New("Unkown fromServer")
```
Comment (nikkiki) [inline main.go]: I would definitely suggest fixing the spelling of `Unknown` here. Is there a better error message we can emit that doesn't use an internal variable name `fromServer`?

File: main.go [hunk 3]
```diff
@@ -234,3 +272,30 @@ func resolveTemplate(templateValue string, event *corev2.Event) (string, error)
...
+func LoginAuth(username, password string) smtp.Auth {
+	return &loginAuth{username, password}
+}
+
+func (a *loginAuth) Start(server *smtp.ServerInfo) (string, []byte, error) {
+	return "LOGIN", []byte(a.username), nil
+}
+
+func (a *loginAuth) Next(fromServer []byte, more bool) ([]byte, error) {
+	if more {
```
Comment (nikkiki) [inline main.go]: Could you elaborate on why `more bool` and the `if more` block is necessary?

Comment (nixwiz) [reply main.go]: That can be found [here](https://golang.org/pkg/net/smtp/#Auth)

**Comment by nixwiz** [comment]
@nikkiki I made some changes regarding loginauth vs. loginAuth to avoid some confusion between the argument and the type that is defined.

File: main.go [hunk 4]
```diff
@@ -143,6 +167,9 @@ func checkArgs(_ *corev2.Event) error {
 	if len(config.ToEmail) == 0 {
 		return errors.New("missing destination email address")
 	}
+	if config.Insecure && config.LoginAuth {
+		return fmt.Errorf("--insecure (-i) and --loginauth (-l) flags are mutually exclusive")
```
Comment (nikkiki) [inline main.go]: This line needs to be updated with the new flag. Personally `enableLoginAuth` reads a bit better, but if you prefer `useLoginAuth` I'm good with that as well 😄

Comment (nixwiz) [reply main.go]: Completed on your suggestion.

**Comment by aphix** [comment]
@nixwiz @nikkiki To be clear, I haven't tested the changes yet, but will attempt to build this branch now and give it a shot-

Thanks for making headway on this one =)

**Comment by nikkiki** [comment]
Thanks for volunteering to test @aphix! It's more helpful than you know 😄

**Comment by aphix** [comment]
@nikkiki @nixwiz Confirmed working on Win 7, testing now in Ubuntu-

Note that I'll not be able to regression test the non-loginAuth path that existed previously (hence why I'm rooting for this PR).

Sorry for the delay, I'm no gopher so there was a bit of env setup/learning required-

**Comment by aphix** [comment]
Following up to say it's confirmed working in Ubuntu -- however in both cases, the included `event.json` failed to work without adding a `timestamp` property at the root.

It does appear to be working when given a sensu-go check output containing only the `spec` portion of the event (when viewing an event via `sensuctl event list --format=wrapped-json`).

I'm having some timeout issues currently using it in a real server (rather than manually triggering with a manufactured event), but I think that's likely network ACLs/firewall rules at this point. Will follow-up when I get more info.

**Comment by aphix** [comment]
@nikkiki @nixwiz 

Some more digging... I've rebuilt from source on a machine that currently works via a _very_ old version of sensu (0.26) to send emails using login auth with an old mailer (`sudo sensu-install -p mailer:1.0.0`), and I'm still getting timeouts.

This tells me that it's likely _not_ an ACL/firewall rule, unless I'm drastically misunderstanding something.

It does immediately exit if I enter an invalid hostname, so I gather it's passing that portion.

Is there any way to debug main.go to get a stack trace from the point where it hangs and eventually times out?

Other details that might help:
The 2 systems that hang are: Ubuntu 18.04.2 LTS, Ubuntu 14.04.5 LTS <- old mailer works
The 2 systems that seem fine: Ubuntu 16.04.5 LTS, Win 7

All systems have used a freshly-built copy using version: `go version go1.12.7 linux/amd64`


Lastly, I'm attaching a small shell script which documents the build process, and if you have golang + git installed, you should be able to update the from/to/host/password variables at the top and run it to build and test the plugin (it includes the sample event updated to include a timestamp).

Hope it helps,
Aphix

P.S. I'll likely complete the testing on Monday, as I'll be heading out for the day.

<details>
<summary>buildAndTestEmailHandler.sh</summary>
#!/bin/bash

set -e

_fromEmail='fromEmail@fromDomain.com'
_toEmail='toEmail@toDomain.com'
_emailHost='email.differentDomainThanFromDomain.com'
_password='some password'

_sampleEventWithTimestamp='{"entity": {"entity_class": "agent","system": {"hostname": "webserver01","os": "linux","platform": "centos","platform_family": "rhel","platform_version": "7.4.1708","network": {"interfaces": [{"name": "lo","addresses": ["127.0.0.1/8","::1/128"]},{"name": "enp0s3","mac": "08:00:27:11:ad:d2","addresses": ["10.0.2.15/24","fe80::26a5:54ec:cf0d:9704/64"]},{"name": "enp0s8","mac": "08:00:27:bc:be:60","addresses": ["172.28.128.3/24","fe80::a00:27ff:febc:be60/64"]}]},"arch": "amd64"},"subscriptions": ["testing","entity:webserver01"],"last_seen": 1542667635,"deregister": false,"deregistration": {},"user": "agent","redact": ["password","passwd","pass","api_key","api_token","access_key","secret_key","private_key","secret"],"metadata": {"name": "webserver01","namespace": "default","labels": null,"annotations": null}},"check": {"check_hooks": null,"duration": 1.903135228,"executed": 1522100915,"high_flap_threshold": 0,"history": [{"executed": 1522100315},{"executed": 1522100915}],"command": "http_check.sh http://localhost:80","handlers": ["slack"],"high_flap_threshold": 0,"interval": 20,"low_flap_threshold": 0,"publish": true,"runtime_assets": [],"subscriptions": ["testing"],"proxy_entity_name": "","check_hooks": null,"stdin": false,"ttl": 0,"timeout": 0,"duration": 0.010849143,"executed": 1542667666,"history": [{"status": 1,"executed": 1542667666}],"issued": 1542667666,"output": "example output","state": "failing","status": 1,"total_state_change": 0,"last_ok": 0,"occurrences": 1,"occurrences_watermark": 1,"output_metric_format": "","output_metric_handlers": [],"env_vars": null,"metadata": {"name": "check-nginx","namespace": "default","labels": null,"annotations": null}},"timestamp": 12345}'

go get github.com/nixwiz/sensu-email-handler

cd "$GOPATH/src/github.com/nixwiz/sensu-email-handler"

git checkout 55b6140624fe22dea05f067913a2bd99466757ac # Should match latest ref in here: https://github.com/sensu/sensu-email-handler/pull/23

go build -o sensu-email-handler main.go

echo "${_sampleEventWithTimestamp}" | ./sensu-email-handler \
	-l \
	-f "${_fromEmail}" \
	-t "${_toEmail}" \
	-s "${_emailHost}" \
	-u "${_fromEmail}" \
	-p "${_password}" \
	-H

echo "Now go check the inbox for: ${_toEmail}"

</details>

**Comment by nixwiz** [comment]
@aphix thanks for your working in testing this.  I'm not sure I currently have access to an email provider that would require this authentication type, but I'll do some checking over the weekend and see what I can come up with.

**Comment by nixwiz** [comment]
@cimere, you originally opened the issue for this (and tested my changes).  I have had to resubmit my changes.  Can you test again?  I have a [64 bit Linux binary](https://toddcampbell.net/sensu/sensu-email-handler) already compiled for you to test if you'd prefer not to bother with attempting to compile for yourself.

**Comment by nixwiz** [comment]
My email hosting provider actually does support the AUTH LOGIN method, so I was able to test this and it works for me.

`cat event.json | ./sensu-email-handler -l -s smtpout.secureserver.net -P 80 -t todd@mydomainname.com -f sensu@mydomainname.com`

Screen Shot 2019-07-19 at 8 39 50 PM

**Comment by aphix** [comment]
Thanks for all the work! Any ETA on getting this versioned and released in Bonsai, or at least via GitHub releases?

**Comment by nixwiz** [comment]
@aphix that's a question for @nikkiki

**Comment by nikkiki** [comment]
@aphix @nixwiz [Release 0.2.0](https://github.com/sensu/sensu-email-handler/releases/tag/0.2.0) 😄

**Comment by nixwiz** [comment]
@nikkiki thanks for that...but it doesn't seem to have refreshed in bonsai for some reason.

**Comment by nikkiki** [comment]
@nixwiz there's apparently some lag with when bonsai fetches from github maybe? It's up there now! https://bonsai.sensu.io/assets/sensu/sensu-email-handler

**Pull request merged**